### PR TITLE
V2 stream xcaching final

### PR DESCRIPTION
**Pull request by 0xakk0r0kamui**

### CODE REVIEW

File: chain/getblocks.go [hunk 1]
```diff
@@ -65,17 +68,13 @@ func (g *BlkGetter) listenCommingBlk(ctx context.Context) {
...
-		ok := g.checkWaitingBlk()
-		if !ok {
-			return
-		}
+	logger.Infof("[stream] listen gnewBlk End")
+	ok := g.checkWaitingBlk()
+	if !ok {
+		logger.Infof("[goroutine] listenCommingBlk END")
+		return
 	}
```
Comment (0xbunyip) [inline chain/getblocks.go]: no need to check for ok

File: chain/getblocks.go [hunk 2]
```diff
@@ -92,7 +91,11 @@ func (g *BlkGetter) updateNewHeight() {
...
 }
 
-func (g *BlkGetter) handleBlkRecv(ctx context.Context, req *proto.BlockByHeightRequest, ch chan common.ExpectedBlk) []uint64 {
+func (g *BlkGetter) handleBlkRecv(
+	ctx context.Context,
+	req *proto.BlockByHeightRequest,
+	ch chan common.ExpectedBlk,
+	providers []Provider) []uint64 {
```
Comment (0xbunyip) [inline chain/getblocks.go]: reformat line break

File: chain/getblocks.go [hunk 3]
```diff
@@ -101,6 +104,14 @@ func (g *BlkGetter) handleBlkRecv(ctx context.Context, req *proto.BlockByHeightR
 			missing = append(missing, blk.Height)
 		} else {
 			g.newBlk <- blk
+			go func(providers []Provider, blk common.ExpectedBlk) {
```
Comment (0xbunyip) [inline chain/getblocks.go]: should we create a new goroutine for each block?

File: chain/getblocks.go (hunk 1 above)
Comment (0xakk0r0kamui) [reply chain/getblocks.go]: done

File: chain/getblocks.go (hunk 2 above)
Comment (0xakk0r0kamui) [reply chain/getblocks.go]: done

**Pull request merged**